broker-core/src/main/java/io/zeebe/broker/subscription/message/state/WorkflowInstanceSubscriptionState.java

```diff
@@ -0,0 +1,233 @@
+import org.agrona.ExpandableArrayBuffer;
+import org.agrona.MutableDirectBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.WriteOptions;
+
+public class WorkflowInstanceSubscriptionState implements StateLifecycleListener {
+
+  private static final String COLUMN_FAMILY_PREFIX = "wf-sub-";
+
+  private static final byte[] SUBSCRIPTIONS_COLUMN_FAMILY_NAME =
+      (COLUMN_FAMILY_PREFIX + "by-key").getBytes();
+  private static final byte[] SENT_TIME_COLUMN_FAMILY_NAME =
+      (COLUMN_FAMILY_PREFIX + "by-sent-time").getBytes();
+
+  public static final byte[][] COLUMN_FAMILY_NAMES = {
+    SUBSCRIPTIONS_COLUMN_FAMILY_NAME, SENT_TIME_COLUMN_FAMILY_NAME
+  };
+
+  private ZbRocksDb db;
+  private ColumnFamilyHandle subscriptionColumnFamily;
+  private ColumnFamilyHandle sentTimeColumnFamily;
+
+  private final ExpandableArrayBuffer keyBuffer = new ExpandableArrayBuffer();
+  private final ExpandableArrayBuffer valueBuffer = new ExpandableArrayBuffer();
+  private final UnsafeBuffer iterateKeyBuffer = new UnsafeBuffer(0, 0);
+
+  private final WorkflowInstanceSubscription subscription = new WorkflowInstanceSubscription();
+
+  public static List<byte[]> getColumnFamilyNames() {
+    return Arrays.asList(COLUMN_FAMILY_NAMES);
+  }
+
+  @Override
+  public void onOpened(StateController stateController) {
+    this.db = stateController.getDb();
+
+    subscriptionColumnFamily =
+        stateController.getColumnFamilyHandle(SUBSCRIPTIONS_COLUMN_FAMILY_NAME);
+    sentTimeColumnFamily = stateController.getColumnFamilyHandle(SENT_TIME_COLUMN_FAMILY_NAME);
+  }
+
+  public void put(final WorkflowInstanceSubscription subscription) {
+    try (final WriteOptions options = new WriteOptions();
+        final ZbWriteBatch batch = new ZbWriteBatch()) {
+
+      subscription.write(valueBuffer, 0);
+
+      batch.put(
+          subscriptionColumnFamily,
+          subscription.getElementInstanceKey(),
+          valueBuffer.byteArray(),
+          subscription.getLength());
+
+      final int keyLength = writeSentTimeKey(keyBuffer, subscription);
+      batch.put(
+          sentTimeColumnFamily, keyBuffer.byteArray(), keyLength, EXISTENCE, EXISTENCE.length);
+
+      db.write(options, batch);
+    } catch (RocksDBException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private int writeSentTimeKey(
+      MutableDirectBuffer buffer, final WorkflowInstanceSubscription subscription) {
+    int offset = 0;
+
+    buffer.putLong(offset, subscription.getCommandSentTime(), STATE_BYTE_ORDER);
+    offset += Long.BYTES;
+
+    buffer.putLong(offset, subscription.getElementInstanceKey(), STATE_BYTE_ORDER);
+    offset += Long.BYTES;
+
+    assert (2 * Long.BYTES) == offset
+        : "Offset problem: offset is not equal to expected key length";
+    return offset;
+  }
+
+  private boolean readSubscription(long key, WorkflowInstanceSubscription subscription) {
```

Comment:
Does it make sense to have an method `readValue` in the ZbRocksDb, which takes an columnFamily, Key and BufferReader ? Otherwise I would say we have these kind of methods on different places. It is not necessary to do this now, but would like to discuss this.